backend/matchAlgorithm.js

```diff
@@ -1,40 +1,143 @@
+          const otherUserPrefList = otherUserPreferences[currentOtherUserId];
+          const [userRank, currentUserRank] = [otherUserPrefList.indexOf(userId), otherUserPrefList.indexOf(currentUserId)];
+
+          if (userRank < currentUserRank) {
+            otherUserMatches[currentOtherUserId] = userId;
+            userMatches[userId] = currentOtherUserId;
+            freeUsers.delete(userId);
+            freeUsers.add(currentUserId);
+          }
+        }
       }
     }
+
+    return { userMatches, otherUserMatches };
+  };
+
+
+  const applyCosineSimilarity = async (currentUser, userPreferences, userType) => {
+    let bestMatch = null;
+    let bestScore = -1;
+
+    for (const id of userPreferences) {
+      const [background, dreams] = userType === 'Adopter'
+        ? [currentUser.background, adoptees.find(adoptee => adoptee.id === id)?.dreams || '']
+        : [adopters.find(adopter => adopter.id === id)?.background || '', currentUser.dreams];
+
+      const similarity = await fetchCosineSimilarity(background, dreams);
+      if (similarity.similarity > bestScore) {
+        bestScore = similarity.similarity;
+        bestMatch = id;
+      }
+
+      await prisma.textScore.upsert({
+        where: { adopterId_adopteeId: { adopterId: parseInt(currentUser.id), adopteeId: parseInt(id) } },
+        update: { score: similarity.similarity },
+        create: { adopterId: parseInt(currentUser.id), adopteeId: parseInt(id), score: similarity.similarity }
+      });
+    }
+
+    return userType === 'Adopter'
+      ? { adopter: currentUser, adoptee: adoptees.find(adoptee => adoptee.id === bestMatch) }
+      : { adopter: adopters.find(adopter => adopter.id === bestMatch), adoptee: currentUser };
+  };
+
+  if (userRole === 'Adopter') {
+    const currentAdopter = adopters.find(adopter => adopter.UserId === loggedInUserId);
+    const currentAdopterPrefList = adopterPreferences[currentAdopter.id];
+
+    const rankDifferent = currentAdopterPrefList.length > 1 &&
+      currentAdopterPrefList.some((id, i, arr) => {
+        if (i === 0) return false;
+        return rawAdopterPreferences.find(pref => pref.adopteeId === id && pref.adopterId === currentAdopter.id).rank !==
+          rawAdopterPreferences.find(pref => pref.adopteeId === arr[i - 1] && pref.adopterId === currentAdopter.id).rank;
+      });
+
+    if (rankDifferent) {
+      const freeAdopters = new Set(adopters.map(adopter => adopter.id));
+      const proposerIndex = adopters.reduce((acc, adopter) => ({ ...acc, [adopter.id]: 0 }), {});
+
+      const { userMatches, otherUserMatches } = applyGaleShapley(adopterPreferences, adopteePreferences, freeAdopters, proposerIndex);
+
+      return {
+        adopter: currentAdopter,
+        adoptee: adoptees.find(adoptee => adoptee.id === userMatches[currentAdopter.id])
+      };
+    } else {
+      return await applyCosineSimilarity(currentAdopter, currentAdopterPrefList, 'Adopter');
+    }
   }
```

Comment:
This check if the preferences have different rank or not and the appropriate matching method is used(Gale-Shapley or cosine similarity)